include/swift/AST/DiagnosticsCommon.def

```diff
@@ -54,6 +54,10 @@ NOTE(previous_decldef,none,
 NOTE(brace_stmt_suggest_do,none,
      "did you mean to use a 'do' statement?", ())
 
+WARNING(defer_stmt_at_block_end,none,
+     "'defer' at the end of its scope is redundant and will execute immediately; "
```

Comment:
@jrose-apple wrote an [excellent guide](https://github.com/apple/swift/blob/master/docs/Diagnostics.md) on Swift diagnostic messages. Some points to consider here include using a terse style (known as reduced written register), phrasing as a rule, and matching the tone of other diagnostics. So perhaps:

> 'defer' statement before end of scope always executes immediately; replace with 'do' statement to silence this warning

Reply:
thank you for the pointer to the Diagnostics guide... useful stuff :+1: .. and yes the message you suggest definitely reads better. Will update it in next commit.

Reply:
warning message improved in ffd13662d11fb1930fe4336cb8aef90d66175a35